laiaaa0/assignment4/Map.cpp

```diff
@@ -0,0 +1,106 @@
+#include "Map.h"
+
+int directions_x [] = {0,0,-1,1};
+int directions_y [] = {1,-1,0,0};
+
+Map::Map(const std::vector<std::vector<char> > & tiles) {
+  this->num_rows_ = tiles.size();
+  if (num_rows_ > 0){
+    this->num_columns_ = tiles[0].size();
+  }
+  else {
+    this->num_columns_ = 0;
+  }
+  this-> map_tiles_ = tiles;
+
+}
+
+Map::Map(const std::string & tiles_str) {
+  this->map_tiles_.clear();
+  std::vector<char> current_row;
+  for (int i = 0; i<tiles_str.size(); ++i){
+      if (tiles_str[i] == '-'){
+        this->map_tiles_.push_back(current_row);
+        current_row.clear();
+      }
+      else {
+        if (tiles_str[i]==ISLAND_T || tiles_str[i]==WATER_T)  {
+          current_row.push_back(tiles_str[i]);
+        }
+      }
+  }
+  this->num_rows_ = this->map_tiles_.size();
+  if (this->num_rows_ > 0){
+    this->num_columns_ = this->map_tiles_[0].size();
+  }
+  else {
+    this->num_columns_ = 0;
+  }
+}
+
+bool Map::InRange(int value, int min, int max) {
+    return value>=min && value<=max;
+}
+
+bool Map::IsInMap(int x, int y) {
+    return InRange(x, 0, this->num_rows_-1) && InRange(y, 0, this->num_columns_-1);
+}
+
+
+bool Map::ExpandNode(int x, int y, std::vector<std::vector<bool> > & visited) {
```

Comment:
Minor thing: pointers are a bit better than references for things you will modify, since it makes it clear to the client.
If you see:
`ExpandNode(x, y, visited)`
you don't know what will happen with visited, but if you see:
`ExpandNode(x, y, &visited)`
is clear that visited may be modified.